core/src/main/java/com/google/errorprone/bugpatterns/nullness/RedundantNullCheck.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2025 The Error Prone Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.errorprone.bugpatterns.nullness;
+
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.matchers.Description.NO_MATCH;
+
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.BinaryTreeMatcher;
+import com.google.errorprone.bugpatterns.nullness.NullnessUtils.NullCheck;
+import com.google.errorprone.dataflow.nullnesspropagation.Nullness;
+import com.google.errorprone.dataflow.nullnesspropagation.NullnessAnnotations;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+
+import com.sun.source.tree.BinaryTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+
+import java.util.Optional;
+
+import javax.lang.model.element.ElementKind;
+
+@BugPattern(
+    summary = "Explicit null check on a variable or method call that is not @Nullable within a @NullMarked scope.",
+    severity = WARNING)
+public class RedundantNullCheck extends BugChecker implements BinaryTreeMatcher {
+
+  @Override
+  public Description matchBinary(BinaryTree tree, VisitorState state) {
+    NullCheck nullCheck = NullnessUtils.getNullCheck(tree);
+    if (nullCheck == null) {
+      return NO_MATCH;
+    }
+
+    VarSymbol varSymbol = nullCheck.varSymbolButUsuallyPreferBareIdentifier();
+    if (varSymbol != null) {
+      if (!NullnessUtils.isInNullMarkedScope(varSymbol, state)) {
+        return NO_MATCH;
+      }
+      if (NullnessUtils.isAlreadyAnnotatedNullable(varSymbol)) {
+        return NO_MATCH;
+      }
+
+      VariableTree varDecl = NullnessUtils.findDeclaration(state, varSymbol);
+
+      if ((varSymbol.getKind() == ElementKind.LOCAL_VARIABLE
+          || varSymbol.getKind() == ElementKind.RESOURCE_VARIABLE) && varDecl != null) {
+
+        if (varDecl.getInitializer() == null) {
+          return NO_MATCH;
+        }
+
```

Comment:
I haven't tested this one out yet, but I'm thinking that we'd want to avoid cases in which the variable is assigned a non-null value but then reassigned later. We _again_ could find clever ways to do that more precisely, but I'd again suggest just skipping the case entirely. I'm hoping that this would do it:

```suggestion
        if (!isConsideredFinal(varSymbol)) {
          return NO_MATCH;
        }

```

(I was about to say that it might render the `getInitializer() == null` check unnecessary, but I now think that that's wrong, since it's possible to write `final Foo foo;` and then assign to it later.)